Fixes to filter out undesirable fences

vlkale opened the pull request:
# Change Made

Fix to invocation of extraneous fences and associated prints at the beginning of fence (kokkosp_begin_fence()) and the ending fence (kokkosp_end_fence()) for the Kokkos debugger tool. This is needed to filter out undesirable fences induced by the debugger, as discussed in Kokkos tools github Issue #147. 

If the name of the function (kernel identifier, represented as the uint64_t kID) the debugger tool is profiling is itself a fence, we filter out the application's fence by setting kId to std::numeric_limits<uint64_t>::max(). 



# Sample Output / Demonstration 

Considering the Kokkos stream benchmark using the kokkos tools logger (note this is without PR #153):

## Before change
The following is done on a macosX with clang++ version 14.0.0. 

```
export KOKKOS_TOOLS_LIBS=${KokkosProjectHOME}/kokkos-tools/debugging/kernel-logger/kp_kernel_logger.so; ./stream.exe

-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Example Library Initialized (sequence is 0, version: 20211015)
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 20 iterations.
-------------------------------------------------------------
KokkosP: Allocate<Host> name: a pointer: 0x12ae00040 size: 800000000
KokkosP: Executing fence on device 100663296 with unique execution identifier 0
KokkosP: Kokkos::Tools::invoke_kokkosp_callback: Kokkos Profile Tool Fence
KokkosP: Execution of fence 0 is completed.
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 1
KokkosP: Kokkos::View::initialization [a] via memset
KokkosP: Executing fence on device 100663297 with unique execution identifier 2
KokkosP: HostSpace fence
KokkosP: Execution of fence 2 is completed.
KokkosP: Executing fence on device 100663296 with unique execution identifier 3
KokkosP: Kokkos::Tools::invoke_kokkosp_callback: Kokkos Profile Tool Fence
KokkosP: Execution of fence 3 is completed.
KokkosP: Execution of kernel 1 is completed.
KokkosP: Executing fence on device 100663297 with unique execution identifier 4
KokkosP: Kokkos::Impl::ViewValueFunctor: View init/destroy fence
KokkosP: Execution of fence 4 is completed.

...

Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                62583.32 MB/s
Copy               81260.04 MB/s
Scale              80542.82 MB/s
Add                81323.77 MB/s
Triad              80843.01 MB/s
-------------------------------------------------------------
KokkosP: Executing fence on device 100663296 with unique execution identifier 525
KokkosP:  HostSpace::impl_deallocate before free
KokkosP: Execution of fence 525 is completed.
KokkosP: Deallocate<Host> name: c pointer: 0x2afaf4040 size: 800000000
KokkosP: Executing fence on device 100663296 with unique execution identifier 526
KokkosP:  HostSpace::impl_deallocate before free
KokkosP: Execution of fence 526 is completed.
KokkosP: Deallocate<Host> name: b pointer: 0x280000040 size: 800000000
KokkosP: Executing fence on device 100663296 with unique execution identifier 527
KokkosP:  HostSpace::impl_deallocate before free
KokkosP: Execution of fence 527 is completed.
KokkosP: Deallocate<Host> name: a pointer: 0x107e00040 size: 800000000
KokkosP: Kokkos library finalization called.

```

 ## After change

```
export CMAKE_CXX_STANDARD=14; export KOKKOS_TOOLS_LIBS=/Users/csuadmin/Desktop/ViveksLaptop/wk/wk/code/kokkosProject/kokkos-tools-new/debugging/kernel-logger/kp_kernel_logger.so; ./stream.exe 
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Example Library Initialized (sequence is 0, version: 20211015)
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 20 iterations.
-------------------------------------------------------------
KokkosP: Allocate<Host> name: a pointer: 0x128800040 size: 800000000
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 0
KokkosP: Kokkos::View::initialization [a] via memset
KokkosP: Executing fence on device 100663297 with unique execution identifier 1
KokkosP: HostSpace fence
KokkosP: Execution of fence 1 is completed.
KokkosP: Execution of kernel 0 is completed.

....


Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                62439.84 MB/s
Copy               80293.91 MB/s
Scale              80096.78 MB/s
Add                80883.20 MB/s
Triad              79857.70 MB/s
-------------------------------------------------------------
KokkosP: Executing fence on device 100663296 with unique execution identifier 317
KokkosP: HostSpace::impl_deallocate before free
KokkosP: Execution of fence 317 is completed.
KokkosP: Deallocate<Host> name: c pointer: 0x2afaf4040 size: 800000000
KokkosP: Executing fence on device 100663296 with unique execution identifier 318
KokkosP: HostSpace::impl_deallocate before free
KokkosP: Execution of fence 318 is completed.
KokkosP: Deallocate<Host> name: b pointer: 0x280000040 size: 800000000
KokkosP: Executing fence on device 100663296 with unique execution identifier 319
KokkosP: HostSpace::impl_deallocate before free
KokkosP: Execution of fence 319 is completed.
KokkosP: Deallocate<Host> name: a pointer: 0x128800040 size: 800000000
KokkosP: Kokkos library finalization called. 

```

Note that compared to the "Before" case, in the "After" case: 
1.  the tool induced fences are filtered out 
2. the unique execution identifier ID  ends at 319 rather than 527 

Also, for this test, the performance, i.e., bandwidth numbers at the end of the output, of the second is not significantly different (within measurement error) from that of the first. We expect that in a larger run and a run on a GPU rather than CPU that performance will improve with this fix.

jhux2 committed:
Fix deprecation in roctx connector

dalg24 committed:
Merge pull request #144 from jhux2/fix-roctx-depr-warning

Fix deprecation in roctx connector

dalg24 committed:
Cleanup header includes

dalg24 committed:
[nvprof-connector] Add support for profiling section

dalg24 committed:
Hand formatting

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

crtrott committed:
Merge pull request #145 from dalg24/nvtx_sections_support

NVTX profiling sections support

dalg24 committed:
[roctx-connector] Do not add extra fences

dalg24 committed:
[roctx-connector] add callbacks for Kokkos fences

dalg24 committed:
Merge pull request #146 from dalg24/roctx-connector-do-not-add-fences

[roctx-connector] Add support for fence profiling

vlkale committed:
Fixing init message for library 

Changing from  'KokkosP: Example Library Initialized ..." to  'KokkosP: High Water Mark Library Initialized ...".

vlkale committed:
Replace default example initialization messages (#150)

* Fixing initialization message in Simple Kernel Timer

* Fixing initialization message from Example in Kernel Logger

* Changing initialization message to be meaningful

* Fixed initialization message

* Update profiling/memory-hwm/kp_hwm.cpp

Co-authored-by: Damien L-G <dalg24+github@gmail.com>

Co-authored-by: Christian Trott <crtrott@sandia.gov>
Co-authored-by: Damien L-G <dalg24+github@gmail.com>

vlkale committed:
filter fence kokkosp_begin_fence function

Putting in fix to filter undesirable (repeated) fence in kokkosp_begin_fence function.

vlkale committed:
remove undesirable fence in kokkosp_end_fence 

 Corresponds to fix of remove undesirable fence in kokkosp_begin_fence

crtrott commented on debugging/kernel-logger/kp_kernel_logger.cpp:
```diff
@@ -137,18 +137,27 @@ extern "C" void kokkosp_end_parallel_reduce(const uint64_t kID) {
 }
 
 extern "C" void kokkosp_end_fence(const uint64_t kID) {
+       if(! (kID == std::numeric_limits<uint64_t>::max() )
```
use != instead of !(... == )

crtrott commented on debugging/kernel-logger/kp_kernel_logger.cpp:
```diff
@@ -137,18 +137,27 @@ extern "C" void kokkosp_end_parallel_reduce(const uint64_t kID) {
+        kokkosp_print_region_stack_indent(level);
+
+        printf("%s\n", name);
+		
```
Is it just me or is a closing brace missing here, so that this wouldn't even compile??

vlkale replied:
Yes, looks like older commit from this past week  - fixing

vlkale committed:
Commit kokkosp_end_fence parens  (and fix statement) 

Commit kokkosp_end_fence parens  (and fix if statement condition )

vlkale committed:
Indent names in logger

Add preceding space for each name, i.e., label, of a region printed by the kernel logger.

vlkale committed:
Adding comments and fixing filter string

vlkale committed:
Support search of case-insensitive label of fence 

Support search of case-insensitive label of fence in the kokkos fence callback for the removal of undesirable fences in the application when it is profiled. This is done through replacing std::strstr() with std::stristr().

vlkale committed:
Label change to tool-induced fence 

Change label to tool-induced fence, so that the fence callback detects only the tool-induced fences, which are known to be 'Kokkos Profile Tool Fence'. This is case-sensitive.

crtrott reviewed (changes requested):
Target develop branch for now, we gonna get rid of master. I think this might be also where the other change comes from. We also need to run a tool over the whole thing to do formatting. But lets do that in a separate PR.

vlkale committed:
4 spaces in front of label instead of 1

Putting 4 spaces instead of 1 for better visual output.

crtrott committed:
Merge pull request #155 from vlkale/IndentNamesInLogger

Indent names in logger (against develop branch)
